### PR TITLE
[6/N] (Elastic EP) Recover failed ranks

### PR DESCRIPTION
**Pull request by UNIDY2002**
## Motivation

As a follow-up to #11657, this PR enables SGLang to dynamically add back previously failed processes, recovering the optimal throughput.

The core idea is as follows. When a node fails, the system admin can relaunch the node with an additional flag `--elastic-ep-rejoin`. Meanwhile, the remaining healthy processes continue serving ongoing inference requests and periodically poll the status of the relaunched process. Once the new processes becomes ready, it is able to seamlessly rejoin the existing process group. **With this design, disruption to ongoing inference is minimized.**

```mermaid
flowchart TD
    X[Healthy processes] --> Y[Failed process found]
    Y --> A
    A[Normal Inference Iteration] --> B{Is new process ready?}
    B -- No --> C[Run inference normally]
    C --> A
    B -- Yes --> D[Join new process into process group]
    D --> C

    U[New process] -->E
    E[Process Relaunched] --> F[Setup Python modules, CUDA, etc.]
    F --> H[Load weight & capture CUDA graph]
    H --> D
```

## Modifications

- Add a new server flag `--elastic-ep-rejoin` to mark a relaunched elastic EP rank that should rejoin an existing Mooncake process group.
- Spread the recovered-rank state throughout distributed initialization so the process groups are created in recovery mode when needed.
- Add elastic EP recovery helpers that poll peer readiness and recover all process groups.
- Resynchronize expert placement metadata after recovery so recovered ranks observe the same global expert-location state as healthy ranks.
- Reset elastic EP active-rank state and the EPLB generator after rank recovery so scheduling resumes from a consistent post-rejoin state.

## Accuracy Tests

Manual multi-node recovery test:

1. Start serving on 2 nodes with elastic EP enabled and enough redundant experts to tolerate one node failure.

Node rank 0:
```bash
sglang serve --model-path <model-path> --trust-remote-code --tp 16 --elastic-ep-backend mooncake --mooncake-ib-device <ib-device-list> --moe-a2a-backend mooncake --deepep-mode low_latency --moe-dense-tp-size 1 --enable-dp-lm-head --enable-two-batch-overlap --disable-custom-all-reduce --enable-eplb --ep-num-redundant-experts <num-redundant-experts> --chunked-prefill-size 512 --cuda-graph-max-bs 16 --max-running-requests 512 --mem-fraction-static 0.5 --enable-dp-attention --dp 16 --device cuda --host 127.0.0.1 --dist-init-addr <ip:port> --port 21000 --nnodes 2 --node-rank 0
```

Node rank 1:
```bash
sglang serve --model-path <model-path> --trust-remote-code --tp 16 --elastic-ep-backend mooncake --mooncake-ib-device <ib-device-list> --moe-a2a-backend mooncake --deepep-mode low_latency --moe-dense-tp-size 1 --enable-dp-lm-head --enable-two-batch-overlap --disable-custom-all-reduce --enable-eplb --ep-num-redundant-experts <num-redundant-experts> --chunked-prefill-size 512 --cuda-graph-max-bs 16 --max-running-requests 512 --mem-fraction-static 0.5 --enable-dp-attention --dp 16 --device cuda --host 127.0.0.1 --dist-init-addr <ip:port> --port 21000 --nnodes 2 --node-rank 1
```

2. Confirm the service is healthy and serving requests.

3. Terminate the process on the larger-index node (`--node-rank 1`).

4. Relaunch that same node with exactly the same command line, only adding `--elastic-ep-rejoin`:

```bash
sglang serve --model-path <model-path> --trust-remote-code --tp 16 --elastic-ep-backend mooncake --mooncake-ib-device <ib-device-list> --moe-a2a-backend mooncake --deepep-mode low_latency --moe-dense-tp-size 1 --enable-dp-lm-head --enable-two-batch-overlap --disable-custom-all-reduce --enable-eplb --ep-num-redundant-experts <num-redundant-experts> --chunked-prefill-size 512 --cuda-graph-max-bs 16 --max-running-requests 512 --mem-fraction-static 0.5 --enable-dp-attention --dp 16 --device cuda --host 127.0.0.1 --dist-init-addr <ip:port> --port 21000 --nnodes 2 --node-rank 1 --elastic-ep-rejoin
```

5. Verify the relaunched rank rejoins successfully and the cluster returns to the healthy state.

Expected result:
- The healthy node continues serving while the replacement rank initializes.
- After the relaunched rank becomes ready, the process groups recover and the node rejoins successfully.
- Logs on healthy ranks should show recovery completing, e.g. `recover ranks [...] done`.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

**Review by UNIDY2002 — COMMENTED** [review]
## Response to Review Comments

Thank you for the detailed review. I've addressed the valid concerns and added explanations for the others.

### Comment 1 (Critical) - Recovery synchronization
This is not an issue. The Mooncake EP backend's internal semantics guarantee that all ranks observe consistent peer readiness state through its own synchronization mechanisms. The polling appears local but is actually coordinated by the library. I've added a comment in the code explaining this: `python/sglang/srt/model_executor/model_runner.py:1445`

### Comment 2 (High) - Dictionary iteration order
**Fixed**: `_iter_live_parallel_groups` now sorts groups by `unique_name` to ensure consistent iteration order across ranks during collective operations. See `python/sglang/srt/elastic_ep/elastic_ep.py:106-111`.

### Comment 3 (High) - CPU copy overhead
**Fixed**: Added early return when `active_ranks.all()` and `active_ranks_cpu.all()` to avoid unnecessary tensor copies in the healthy case. See `python/sglang/srt/model_executor/model_runner.py:1435-1436`.

### Comment 4 (Medium) - src_rank=0 safety
The function signature keeps the default for backward compatibility, but I've added a docstring comment explaining the caller's responsibility to ensure the source rank is healthy. In recovery scenarios, this function is called after all ranks have recovered.

### Comment 5 (Medium) - Daemon thread
**Fixed**: Added `daemon=True` to the background thread. See `python/sglang/srt/managers/data_parallel_controller.py:356`.

### Comment 6 (Medium) - ranks[0] for broadcast
No change needed. At this point in the recovery flow, all ranks (including rank 0) have successfully recovered, so using `ranks[0]` as the source is correct. Changing this could cause inconsistency with `tp_worker`'s `broadcast_pyobj` synchronization.

File: python/sglang/srt/model_executor/model_runner.py [hunk 1]
```diff
@@ -1391,6 +1407,29 @@ def update_expert_location(
...
 
+    def maybe_recover_ep_ranks(self):
+        tp_active_ranks = self.tp_group.active_ranks.detach().cpu().numpy()
+        tp_active_ranks_cpu = self.tp_group.active_ranks_cpu.detach().numpy()
+        tp_active_ranks &= tp_active_ranks_cpu
+        ranks_to_recover = [
+            i for i in range(len(tp_active_ranks)) if not tp_active_ranks[i]
+        ]
+
+        if ranks_to_recover and try_recover_ranks(ranks_to_recover):
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py:1484]: This is not an issue. The Mooncake EP backend's internal semantics guarantee that all ranks observe consistent peer readiness state through its own synchronization mechanisms. The polling appears local but is actually coordinated by the library. I've added a comment in the code explaining this: python/sglang/srt/model_executor/model_runner.py:1445

File: python/sglang/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -1415,6 +1431,35 @@ def update_expert_location(
...
+        # are safe even though polling appears local.
+        if ranks_to_recover and try_recover_ranks(ranks_to_recover):
+            self.forward_pass_id = 0
+            self.eplb_manager.reset_generator()
+            broadcast_global_expert_location_metadata()
+            ElasticEPStateManager.instance().reset()
+
+            broadcast_pyobj(
+                [self.server_args.random_seed],
+                self.tp_size * self.pp_rank + self.tp_rank,
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py]: **Fixed**: Changed from `self.tp_size * self.pp_rank + self.tp_rank` to `get_world_group().rank` to correctly obtain the global rank. The previous calculation was incorrect when DP is enabled since multiple DP replicas would produce the same value. See commit 446a75934.

File: python/sglang/srt/eplb/expert_location.py [hunk 3]
```diff
@@ -319,6 +319,52 @@ def set_global_expert_location_metadata(value):
     _global_expert_location_metadata = value
 
 
+def broadcast_global_expert_location_metadata(
+    src_rank: int = 0, group: Optional[torch.distributed.ProcessGroup] = None
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/eplb/expert_location.py:322]: **Fixed**: The recovery code now dynamically selects a healthy rank (one not in `ranks_to_recover`) as the broadcast source instead of defaulting to rank 0. This prevents broadcasting stale metadata when rank 0 is the one being recovered. The healthy source rank is computed as `self.tp_group.ranks[healthy_local_indices[0]]`. See commit 446a75934.

File: python/sglang/srt/managers/data_parallel_controller.py [hunk 4]
```diff
@@ -351,6 +351,25 @@ def _broadcast_ports_as_server(
             return worker_ports
         finally:
             rep_socket.close()
+            if self.server_args.elastic_ep_backend is not None:
+                threading.Thread(
+                    target=self._reply_ports_as_server, args=(endpoint, worker_ports),
+                    daemon=True,
+                ).start()
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/managers/data_parallel_controller.py:366]: Acknowledged. This is a valid concern, but the ZMQ thread lifecycle management is not critical for the current PR scope. The daemon thread was already added in the previous fix (commit a16025b0d), and the re-initialization scenario is unlikely in typical usage. We can address this in a follow-up if needed.

File: python/sglang/srt/model_executor/model_runner.py [hunk 5]
```diff
@@ -1415,6 +1431,43 @@ def update_expert_location(
                     weight_name_filter=weight_name_filter,
                 )
 
+    def maybe_recover_ep_ranks(self):
+        if self.tp_group.active_ranks.all() and self.tp_group.active_ranks_cpu.all():
+            return
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py:1468]: This logic happens between forward iterations, and this syncing should have little impact on the overall performance, so overlooking it by now.

File: python/sglang/srt/model_executor/model_runner.py [hunk 6]
```diff
@@ -1415,6 +1431,43 @@ def update_expert_location(
...
+            healthy_local_indices = [
+                i for i in range(len(tp_active_ranks)) if i not in ranks_to_recover
+            ]
+            healthy_src_rank = (
+                self.tp_group.ranks[healthy_local_indices[0]]
+                if healthy_local_indices
+                else get_world_group().ranks[0]
+            )
+            broadcast_global_expert_location_metadata(src_rank=healthy_src_rank)
+            ElasticEPStateManager.instance().reset()
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py:1492]: The underlying mooncake library's semantics say that the active_ranks are synced to all 1 after connecting.

File: python/sglang/srt/elastic_ep/elastic_ep.py [hunk 7]
```diff
@@ -71,3 +89,115 @@ def healthy_rank_state(
...
+
+def _maybe_create_message_queue(group) -> None:
+    if not group.use_message_queue_broadcaster or group.world_size <= 1:
+        return
+
+    from sglang.srt.distributed.device_communicators.shm_broadcast import MessageQueue
+
+    group.mq_broadcaster = MessageQueue.create_from_process_group(
+        group.cpu_group, 1 << 22, 6
+    )
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/elastic_ep/elastic_ep.py:138]: False positive

File: python/sglang/srt/model_executor/model_runner.py [hunk 8]
```diff
@@ -1415,6 +1431,43 @@ def update_expert_location(
                     weight_name_filter=weight_name_filter,
                 )
 
+    def maybe_recover_ep_ranks(self):
+        if self.tp_group.active_ranks.all() and self.tp_group.active_ranks_cpu.all():
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py:1467]: False positive. Won't fix.

File: python/sglang/srt/managers/data_parallel_controller.py [hunk 9]
```diff
@@ -351,6 +351,26 @@ def _broadcast_ports_as_server(
...
+                    target=self._reply_ports_as_server,
+                    args=(endpoint, worker_ports),
+                    daemon=True,
+                ).start()
+
+    def _reply_ports_as_server(self, endpoint: str, worker_ports: List[int]):
+        """
+        Runs as a background thread to broadcast worker ports for recovered EP ranks
+        """
+        rep_socket = get_zmq_socket(self.context, zmq.REP, endpoint, True)
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/managers/data_parallel_controller.py]: Fixed.

File: python/sglang/srt/model_executor/model_runner.py [hunk 10]
```diff
@@ -1415,6 +1431,43 @@ def update_expert_location(
...
+            self.forward_pass_id = 0
+            self.eplb_manager.reset_generator()
+            healthy_local_indices = [
+                i for i in range(len(tp_active_ranks)) if i not in ranks_to_recover
+            ]
+            healthy_src_rank = (
+                self.tp_group.ranks[healthy_local_indices[0]]
+                if healthy_local_indices
+                else get_world_group().ranks[0]
+            )
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py]: Fixed.

File: python/sglang/srt/model_executor/model_runner.py [hunk 11]
```diff
@@ -478,6 +484,14 @@ def __init__(
         self.initialize(pre_model_load_memory)
         self.check_quantized_moe_compatibility()
 
+        if (
+            self.server_args.elastic_ep_backend is not None
+            and self.server_args.elastic_ep_rejoin
+        ):
+            join_process_groups()
+            broadcast_global_expert_location_metadata()
+            ElasticEPStateManager.instance().reset()
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py:499]: Fixed.

File: python/sglang/srt/model_executor/model_runner.py [hunk 12]
```diff
@@ -1438,6 +1456,51 @@ def update_expert_location(
                     weight_name_filter=weight_name_filter,
                 )
 
+    def maybe_recover_ep_ranks(self):
+        if self.tp_group.active_ranks.all() and self.tp_group.active_ranks_cpu.all():
+            return
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py:1468]: Leaving to future work. It only impacts Elastic EP path, and e2e tests show such performance impact is minimal.

File: python/sglang/srt/model_executor/model_runner.py [hunk 13]
```diff
@@ -480,6 +486,18 @@ def __init__(
...
+            self.server_args.elastic_ep_backend is not None
+            and self.server_args.elastic_ep_rejoin
+        ):
+            join_process_groups()
+            broadcast_global_expert_location_metadata(
+                src_rank=self._get_healthy_expert_location_src_rank(
+                    invoked_in_elastic_ep_rejoin_path=True
+                )
+            )
+            ElasticEPStateManager.instance().reset()
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py:499]: False positive. The random-seed broadcast happens after ModelRunner initialization, so no need to execute it here.

File: python/sglang/srt/model_executor/model_runner.py [hunk 14]
```diff
@@ -1438,6 +1458,66 @@ def update_expert_location(
...
+
+            broadcast_pyobj(
+                [self.server_args.random_seed],
+                get_world_group().rank,
+                get_world_group().cpu_group,
+                src=get_world_group().ranks[0],
+            )
```
[the remark this answers is not in the document]
Comment (UNIDY2002) [reply python/sglang/srt/model_executor/model_runner.py:1499]: False positive.

**Comment by UNIDY2002** [comment]
This PR mainly brings incremental changes, and the logics are guarded with conditions so that non-elastic-ep paths will not be affected. There remains a minor performance issue regarding elastic EP recovery that will be addressed in follow-up PRs.

cc @ShangmingCai

File: python/sglang/srt/model_executor/model_runner.py [hunk 15]
```diff
@@ -1438,6 +1458,66 @@ def update_expert_location(
...
+        # synchronization overhead does not propagate to other configs.
+        # Leave for future optimization of the elastic EP path.
+        if self.tp_group.active_ranks.all() and self.tp_group.active_ranks_cpu.all():
+            return
+
+        tp_active_ranks = self.tp_group.active_ranks.detach().cpu().numpy()
+        tp_active_ranks_cpu = self.tp_group.active_ranks_cpu.detach().numpy()
+        tp_active_ranks &= tp_active_ranks_cpu
+        # NOTE: `ranks_to_recover` uses indices in `tp_group`. For the current
+        # Mooncake elastic EP implementation we assume `--pp-size=1`, so the
```
Comment (ShangmingCai) [inline python/sglang/srt/model_executor/model_runner.py:1470]: Maybe add an assertion in server_args.py when pp size > 1

File: python/sglang/srt/distributed/parallel_state.py
Comment (ShangmingCai) [inline python/sglang/srt/distributed/parallel_state.py:1]: CC: @yizhang2077

**Review by ShangmingCai — COMMENTED** [review]
Looks good. cc: @yizhang2077 @ch-wan

**Comment by UNIDY2002** [comment]
截屏2026-04-23 09 14 28

**Comment by UNIDY2002** [comment]
Wait, we have a new conflict.

**Comment by UNIDY2002** [comment]
Current CI errors arise from irrelevant upstream regressions.

**Review by ch-wan — APPROVED** [review]
